test/pynrn/test_fast_imem.py

```diff
@@ -0,0 +1,192 @@
+# Sum of all i_membrane_ should equal sum of all ElectrodeCurrent
+# For a demanding test, use a tree with many IClamp and ExpSyn point processes
+# sprinkled on zero and non-zero area nodes.
+
+from neuron import h
+
+class Cell():
+  def __init__(self, id, nsec):
+    self.id = id
+    self.secs = [h.Section(name="d"+str(i), cell=self) for i in range(nsec)]
+    r = h.Random()
+    r.Random123(id, 0, 0)
+
+    # somewhat random tree, d[0] plays role of soma with connections to
+    # d[0](0.5) and all others to 1.0
+    for i in range(1, nsec):
+      iparent = int(r.discunif(0, i-1))
+      x = 0.5 if iparent == 0 else 1.0
+      self.secs[i].connect(self.secs[iparent](x))
+
+    # uniform L and diam but somewhat random passive g and e
+    for i, sec in enumerate(self.secs):
+      sec.L = 10 if i > 0 else 5
+      sec.diam = 1 if i > 0 else 5
+      sec.insert("pas")
+      sec.g_pas = 0.0001*r.uniform(1.0, 1.1)
+      sec.e_pas = -65*r.uniform(1.0, 1.1)
+
+    # IClamp and ExpSyn at every location (even duplicates) with random
+    # parameters (would rather use a Shunt, but ...)
+    self.ics = []
+    self.syns = []
+    self.netcons = []
+    self.netstim = h.NetStim()
+    self.netstim.number=1
+    self.netstim.start=0.0
+    for sec in self.secs:
+      for seg in sec.allseg():
+        ic = h.IClamp(seg)
+        ic.delay = 0.1
+        ic.dur = 1.0
+        ic.amp = .001*r.uniform(1.0, 1.1)
+        self.ics.append(ic)
+
+        syn = h.ExpSyn(seg)
+        syn.e = -65*r.uniform(1.0, 1.1)
+        syn.tau = r.uniform(0.1, 1.0)
+        self.syns.append(syn)
+        nc = h.NetCon(self.netstim, syn)
+        nc.delay = 0.1
+        nc.weight[0] = 0.001*r.uniform(1.0, 1.1)        
+        self.netcons.append(nc)
+
+  def __str__(self):
+    return "Cell"+str(self.id)
+
+def total_imem():
+  imem = 0.0
+  for sec in h.allsec():
+    for seg in sec.allseg():
+      if seg.x == 0.0 and sec.parentseg() != None:
+        assert(seg.i_membrane_ == sec.parentseg().i_membrane_)
+        continue # don't count twice
+      imem += seg.i_membrane_
+  #print("total_imem ", imem)
+  return imem
+	
+def total_iclamp(ics):
+  icur = 0.0
+  for ic in ics:
+    icur += ic.i
+  #print("total_iclamp ", icur)
+  return icur
+
+# verified that events arrived at syn to change g.
+def total_syn_g(syns):
+  g = 0
+  for syn in syns:
+    g += syn.g
+  #print("total syn g ", g)
+  return g
+
+#Remark:
+#  The iteration over all segments (including 0 area nodes) without
+#  counting 0 area nodes twice is
+'''
+    for sec in h.allsec():
+     for seg in sec.allseg():
+       if seg.x == sec.orientation() and sec.parentseg() != None:
+         continue
+       # rest of the "for seg..." body
+'''
+#  Note that sec.parentseg() is needed to count the root and use of
+#  sec.trueparentseg() would miss the root node. Also, although an
+#  extremely rare edge case, sec.orientation() is needed to match
+#  which segment is closest to root.
+
```

Comment:
is this very rare edge case tested here?

Reply:
There was an attempt to do so with the ```test_allseg_unique_iter()``` where the orientation of sections d, e, g, g1, and g2
is nonstandard. Do you think it is worthwhile to do some more explicit testing in the context of summing ```i_membrane_``` by also using the ```test_allseg_unique_iter()``` topology (i.e. basically a full test analog to the ```class Cell```

Reply:
probably yes. However, I have never seen this case happening in real life